### PR TITLE
Store: Add the ability to update a review's status and delete a review.

### PR DESCRIPTION
**Pull request by justinshreve**
This PR adds the code necessary to manage a review's status, including permanently deleting it from the 'trash' queue. 

This is based on the design from #14177 with some differences to bring the experience inline with the new comments interface, which this PR is patterned off of.

To Test:
* Make sure you are the latest master of `wc-api-dev` (API changes have been merged). (I'll be getting this deployed for AT sites shortly).
* Run `npm run test-client client/extensions/woocommerce/state` and make sure all tests pass.
* Make sure you have a few reviews, including at least 1 pending review.
* Go to `http://calypso.localhost:3000/store/reviews/:site`.
* Verify that you can see your pending review. Expand it to open up the action bar, and select "Approve".
* Verify that your review approves. You should see a success message and the entry should change colors.
* Open the review again and trash it. You should again see a success message, and it should leave the queue and go into the trash tab.
* Go to the 'Trash' queue, and find the review. Click 'Delete' to actually delete the review.

Depends on a fix from #18168.

### CODE REVIEW

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 1]
```diff
@@ -62,3 +79,99 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+
+export function announceDeleteSuccess( { dispatch, getState }, action ) {
+	const state = getState();
+	const { siteId } = action;
+	const currentPage = getReviewsCurrentPage( state, siteId );
+	dispatch( fetchReviews( siteId, { search: '', page: currentPage, status: 'trash' } ) );
+
+	dispatch(
+		successNotice(
+			translate( 'Review deleted permanently.' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'Comment deleted permanently.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 2]
```diff
@@ -62,3 +79,99 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+	dispatch(
+		bypassDataLayer( {
+			...omit( action, [ 'meta' ] ),
+			newStatus: currentStatus,
+		} )
+	);
+
+	const errorMessage = {
+		approved: translate( "We couldn't approve this review." ),
+		pending: translate( "We couldn't unapprove this review." ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'We couldn't unapprove this comment.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 3]
```diff
@@ -62,3 +79,99 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+			query.status = 'any';
+		}
+
+		dispatch( fetchReviews( siteId, query ) );
+	}
+
+	const message = {
+		approved: translate( 'Review approved.' ),
+		pending: translate( 'Review unapproved.' ),
+		spam: translate( 'Review marked as spam.' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Comment marked as spam.' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 4]
```diff
@@ -62,3 +79,99 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+		if ( '' !== currentSearch ) {
+			query.status = 'any';
+		}
+
+		dispatch( fetchReviews( siteId, query ) );
+	}
+
+	const message = {
+		approved: translate( 'Review approved.' ),
+		pending: translate( 'Review unapproved.' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Hi! I've found a possible matching string that has already been translated **103** times:
`translate( 'Unapproved' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 5]
```diff
@@ -62,3 +79,99 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+
+		if ( '' !== currentSearch ) {
+			query.status = 'any';
+		}
+
+		dispatch( fetchReviews( siteId, query ) );
+	}
+
+	const message = {
+		approved: translate( 'Review approved.' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Hi! I've found a possible matching string that has already been translated **85** times:
`translate( 'Approved' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/reviews/review-actions-bar.js [hunk 6]
```diff
@@ -11,11 +12,21 @@ import PropTypes from 'prop-types';
...
  */
 import Button from 'components/button';
+import { changeReviewStatus, deleteReview } from 'woocommerce/state/sites/reviews/actions';
 
 // currentStatus is the status/tab that is currently selected in the UI.
 // This is used to show the 'delete' action on the trash page.
 // review.status is the status of the individual review.
-const ReviewActionsBar = ( { review, currentStatus, translate } ) => {
+const ReviewActionsBar = ( {
+	review,
```
Comment (allendav) [inline client/extensions/woocommerce/app/reviews/review-actions-bar.js]: Would you mind alphabetizing these? Makes it easier to find things later

File: client/extensions/woocommerce/app/reviews/review-actions-bar.js [hunk 7]
```diff
@@ -69,10 +83,40 @@ const ReviewActionsBar = ( { review, currentStatus, translate } ) => {
...
 	review: PropTypes.shape( {
 		status: PropTypes.string,
 	} ).isRequired,
 	currentStatus: PropTypes.string.isRequired,
+	toggleExpanded: PropTypes.func,
 };
 
-export default localize( ReviewActionsBar );
+const mapDispatchToProps = ( dispatch, ownProps ) => {
+	const { review, review: { product }, siteId, currentStatus, toggleExpanded } = ownProps;
```
Comment (allendav) [inline client/extensions/woocommerce/app/reviews/review-actions-bar.js]: I carefully looked through mozilla's destructuring assignment page, but i couldn't find an example where the same key was used twice like you did here for review. Are you trying to merge fields into review? I did a little experimenting in the javascript console and was unable to get merging to happen, i.e.:

screen shot 2017-09-22 at 9 43 26 am

Comment (justinshreve) [reply client/extensions/woocommerce/app/reviews/review-actions-bar.js]: It's to pull another object out of review itself, but I agree it's not clear. I can change this.

File: client/extensions/woocommerce/state/sites/reviews/handlers.js [hunk 8]
```diff
@@ -62,3 +79,100 @@ export function handleReviewsRequestError( { dispatch }, action, error ) {
...
+			translate( 'Review deleted permanently.' ),
+			{ duration: 5000 }
+		)
+	);
+}
+
+export function announceDeleteFailure( { dispatch } ) {
+	dispatch(
+		errorNotice(
+			translate( 'Could not delete the review.' ),
```
Comment (allendav) [inline client/extensions/woocommerce/state/sites/reviews/handlers.js]: Above you used a "We couldn't..." pattern, but here the error is plainer - lets make them consistent one way or the other.

**Comment by allendav** [comment]
Toggling a review back and forth between approved (and not) doesn't seem to update the current list view (a hard refresh is needed)

approve-unapprove-approve

**Pull request merged**